pkg/solana/logpoller/parser.go

```diff
@@ -1,5 +1,413 @@
+			if err != nil {
+				return "", err
+			}
+
+			if clause != "" {
+				clauses = append(clauses, clause)
+			}
+		} else {
+			clause, err := v.combineExpressions(exp.BoolExpression.Expressions, exp.BoolExpression.BoolOperator)
+			if err != nil {
+				return "", err
+			}
+
+			if clause != "" {
+				clauses = append(clauses, clause)
+			}
+		}
+	}
+
+	if len(clauses) == 0 {
+		return "", nil
+	}
+
+	output := strings.Join(clauses, fmt.Sprintf(" %s ", op.String()))
+
+	if len(clauses) > 1 {
+		output = fmt.Sprintf("(%s)", output)
+	}
+
+	return output, nil
+}
+
+func (v *pgDSLParser) getLastExpression() (string, error) {
+	exp := v.expression
+	err := v.err
+
+	v.expression = ""
+	v.err = nil
+
+	return exp, err
+}
+
+func cmpOpToString(op primitives.ComparisonOperator) (string, error) {
+	switch op {
+	case primitives.Eq:
+		return "=", nil
+	case primitives.Neq:
+		return "!=", nil
+	case primitives.Gt:
+		return ">", nil
+	case primitives.Gte:
+		return ">=", nil
+	case primitives.Lt:
+		return "<", nil
+	case primitives.Lte:
+		return "<=", nil
+	default:
+		return "", ErrInvalidComparator
+	}
+}
+
+// ensure valuesFromCursor remains consistent with the function above that creates a cursor
+func valuesFromCursor(cursor string) (int64, int, []byte, error) {
+	partCount := 3
+
+	parts := strings.Split(cursor, "-")
+	if len(parts) != partCount {
+		return 0, 0, nil, fmt.Errorf("%w: must be composed as block-logindex-txHash", ErrInvalidCursorFormat)
+	}
+
+	block, err := strconv.ParseInt(parts[0], 10, 64)
+	if err != nil {
+		return 0, 0, nil, fmt.Errorf("%w: block number not parsable as int64", ErrInvalidCursorFormat)
+	}
+
+	logIdx, err := strconv.ParseInt(parts[1], 10, 32)
+	if err != nil {
+		return 0, 0, nil, fmt.Errorf("%w: log index not parsable as int", ErrInvalidCursorFormat)
+	}
+
+	txHash, err := solana.PublicKeyFromBase58(parts[2])
+	if err != nil {
+		return 0, 0, nil, fmt.Errorf("%w: invalid transaction hash: %s", ErrInvalidCursorFormat, err.Error())
+	}
+
+	return block, int(logIdx), txHash.Bytes(), nil
+}
+
+func orderToString(dir query.SortDirection) (string, error) {
+	switch dir {
+	case query.Asc:
+		return "ASC", nil
+	case query.Desc:
+		return "DESC", nil
+	default:
+		return "", ErrInvalidSortDir
+	}
+}
+
+type addressFilter struct {
+	address solana.PublicKey
+}
+
+func NewAddressFilter(address solana.PublicKey) query.Expression {
+	return query.Expression{
+		Primitive: &addressFilter{address: address},
+	}
+}
+
+func (f *addressFilter) Accept(visitor primitives.Visitor) {
+	switch v := visitor.(type) {
+	case *pgDSLParser:
+		v.VisitAddressFilter(f)
+	}
+}
+
+type eventSigFilter struct {
+	eventSig []byte
+}
+
+func NewEventSigFilter(sig []byte) query.Expression {
+	return query.Expression{
+		Primitive: &eventSigFilter{eventSig: sig},
+	}
+}
+
+func (f *eventSigFilter) Accept(visitor primitives.Visitor) {
+	switch v := visitor.(type) {
+	case *pgDSLParser:
+		v.VisitEventSigFilter(f)
+	}
+}
```

Comment:
I think we're missing an essential feature here that CCIP will require, and that we won't be able to implement QueryKey without. You have `eventSigFilter`, `eventByAddressFitler`, but we also need an `eventBySubKeyFilter` so that we can filter by subkey.

We'll need that to implement support for `IndexedSequencesKeyFilter` and several other `IndexedSequences...` filters that can be passed to `QueryKey`. CCIP definitely uses some of those.

Reply:
See:  `func (b *EventBinding) encodeComparator(comparator *primitives.Comparator) (query.Expression, error)` in `chainlink/core/services/relay/evm/read/event.go`)

We'll have to call `NewEventBySubKey` there instead of `NewEventByTopicFilter` and `NewEventBySubKeyFilter`.

Reply:
I will be adding this when I do the remapping.